src/itwinai/slurm/slurm_script_builder.py

```diff
@@ -0,0 +1,299 @@
+            self.omp_num_threads = self.cpus_per_gpu
+        else:
+            self.omp_num_threads = 1
+
+    def update_autogenerated_fields(self):
+        self.job_name = self.generate_identifier()
+        self.std_out = Path(f"slurm_jobs/{self.generate_identifier()}.out")
+        self.err_out = Path(f"slurm_jobs/{self.generate_identifier()}.err")
+
+    def generate_identifier(self) -> str:
+        return f"{self.distributed_strategy}-{self.num_nodes}x{self.gpus_per_node}"
+
+    def get_training_command(self) -> str:
+        return "python main.py"
+
+    def get_srun_command(self, torch_log_dir: str = "logs_torchrun"):
+        if self.distributed_strategy in ["ddp", "deepspeed"]:
+            rdzv_endpoint = (
+                '\'$(scontrol show hostnames "$SLURM_JOB_NODELIST"'
+                " | head -n 1)':29500"
+            )
+            main_command = rf"""
+            srun --cpu-bind=none --ntasks-per-node=1 \
+            bash -c "torchrun \
+                --log_dir='{torch_log_dir}' \
+                --nnodes=$SLURM_NNODES \
+                --nproc_per_node=$SLURM_GPUS_PER_NODE \
+                --rdzv_id=$SLURM_JOB_ID \
+                --rdzv_conf=is_host=\$(((SLURM_NODEID)) && echo 0 || echo 1) \
+                --rdzv_backend=c10d \
+                --rdzv_endpoint={rdzv_endpoint} \
+                {self.get_training_command()}"
+        """
+        else:
+            num_tasks = self.gpus_per_node * self.num_nodes
+
+            # E.g. "0,1,2,3" with 4 GPUs per node
+            cuda_visible_devices = ",".join(str(i) for i in range(self.gpus_per_node))
+
+            main_command = rf"""
+            export CUDA_VISIBLE_DEVICES="{cuda_visible_devices}"
+            srun --cpu-bind=none \
+                --ntasks-per-node=$SLURM_GPUS_PER_NODE \
+                --cpus-per-task=$SLURM_CPUS_PER_GPU \
+                --ntasks={num_tasks} \
+                {self.get_training_command()}
+        """
+
+        main_command = main_command.strip()
+        return remove_indentation_from_multiline_string(main_command)
+
+    def get_debug_command(self):
+        debug_print_command = rf"""
+        echo "### SLURM Variables ###"
+        echo "SLURM_SUBMIT_DIR: $SLURM_SUBMIT_DIR"
+        echo "SLURM_JOB_ID: $SLURM_JOB_ID"
+        echo "SLURM_JOB_NODELIST: $SLURM_JOB_NODELIST"
+        echo "SLURM_NNODES: $SLURM_NNODES"
+        echo "SLURM_NTASKS: $SLURM_NTASKS"
+        echo "SLURM_TASKS_PER_NODE: $SLURM_TASKS_PER_NODE"
+        echo "SLURM_SUBMIT_HOST: $SLURM_SUBMIT_HOST"
+        echo "SLURMD_NODENAME: $SLURMD_NODENAME"
+        echo "CUDA_VISIBLE_DEVICES: $CUDA_VISIBLE_DEVICES"
+        echo ""
+        echo "### Other Variables ###"
+        echo "Distributed Strategy: {self.distributed_strategy}"
+        """
+        debug_print_command = debug_print_command.strip()
+        return remove_indentation_from_multiline_string(debug_print_command)
+
+    def get_setup_command(self):
```

Comment:
I would rename setup as pre_exec for better clarity, and possibly add a post_exec to be run after the srun (e.g., cleanup)